### DISCUSSION TITLE
Use of subset_fonts results in loss of part of the page.

### DISCUSSION THREAD

**Discussion by izerui** [pymupdf/PyMuPDF#3368] [Looking for help] [answered]
```
def test_subset_fonts():
    doc = fitz.open('401-020605-00.pdf')
    doc.subset_fonts()
    doc.save('x.pdf')
```

![1](https://github.com/pymupdf/PyMuPDF/assets/1773233/33ca0cec-6258-480f-9893-12c1f0d9f613)
![2](https://github.com/pymupdf/PyMuPDF/assets/1773233/4f1deea0-96fc-4769-8c92-faa3281d4b09)

pdf:
[401-020605-00.pdf](https://github.com/pymupdf/PyMuPDF/files/14926684/401-020605-00.pdf)

**Comment by JorjMcKie** [comment, marked as answer]
This method is based on an external package, fontTools, which is not error-free.
So we are not in the position to track down the problem.
In addition we are phasing out this method - another reason why we do not pursue errors here.

We are in the process to publish our own font sub-setter, which is directly implemented in MuPDF. It is very much faster (about 15 times) and covers a larger range of fonts.
It is available as a function in `fitz.mupdf`, but is yet experimental and has no elegant wrapping in PyMuPDF's API yet. It will not report anything back, i.e. there is no "verbose" option or similar.
Here is how you can use it today. I tried your example and obviously MuPDF's font sub-setter determined that the 4 used fonts cannot be subset.
```python
import fitz
doc = fitz.open("input.pdf")
pdf = fitz._as_pdf_document(doc)  # access underlying PDF document of the general Document
fitz.mupdf.pdf_subset_fonts2(pdf, list(range(doc.page_count)))  # create font subsets
```

As you can see, these are just two lines, but the use still is a little clumsy.

**Reply by izerui** [reply]
Ok, thanks a lot, looking forward to your font sub-setter.@_@

**Reply by izerui** [reply]
At the risk of being presumptuous, what is the approximate release date for this feature.

**Reply by JorjMcKie** [reply]
no date fixed yet - but as mentioned: go ahead and use those two lines today.

**Comment by izerui** [comment]
Perhaps I have a suggestion, you can refer to, in the page to add text, whether we can directly use an external font library and not embedded in the pdf, I can choose a majority of our country's operating system will exist in a font to render, so that in that can ensure that the volume of the pdf can also be a normal display.

**Reply by JorjMcKie** [reply]
This is not supported by MuPDF. All fonts must either be Base-14 (Helvetica, Times-Roman, Courier), or they must be explicitly provided as font files.
There never happens searching the font library of the OS or whatever.

**Reply by izerui** [reply]
Okay, thanks for your patience.
